nbs_probabilistic/create_dropout_ensemble.py

```diff
@@ -0,0 +1,131 @@
+    #changing paths
+#   model_save_dir='/home/garg/data/WeatherBench/predictions/saved_models'
+#   datadir='/home/garg/data/WeatherBench/5.625deg'
+    
+    #run data generator
+    ds = xr.merge([xr.open_mfdataset(f'{datadir}/{var}/*.nc', combine='by_coords') 
+                   for var in var_dict.keys()])
+    mean = xr.open_dataarray(f'{model_save_dir}/{exp_id}_mean.nc') 
+    std = xr.open_dataarray(f'{model_save_dir}/{exp_id}_std.nc')
+    
+    ds_test= ds.sel(time=slice(test_years[0],test_years[-1]))
+    #Ques: Should shuffle be False? since its testing.
+    dg_test = DataGenerator(ds_test, var_dict, lead_time, batch_size=batch_size, shuffle=False,
+                            load=True,mean=mean, std=std, output_vars=output_vars,
+                            data_subsample=data_subsample, norm_subsample=norm_subsample,
+                            nt_in=nt_in,dt_in=dt_in) 
+    return dg_test
+
+def get_model(exp_id, model_save_dir):
+    tf.compat.v1.disable_eager_execution() #needed
+    saved_model_path=f'{model_save_dir}/{exp_id}.h5'
+    
+    substr=['resnet','unet_google','unet']
+    assert any(x in exp_id for x in substr)
+    model=tf.keras.models.load_model(saved_model_path,
+                                     custom_objects={'PeriodicConv2D':PeriodicConv2D,
+                                    'lat_mse': tf.keras.losses.mse})
+    #ToDo: add other loss functions to custom_objects. doesn't matter if it is not used in the model itself, only so that load_model() doesn't break)
+    #Since we dont build again, we dont need to pass model params like kernel, filters, activation, dropout,loss and other details to the network.
+
+    #model.summary()
+    return model
+
+def predict(dg_test,model,number_of_forecasts, output_vars):
+    #For just 1 batch of data
+#     X,y=dg_test[0] #currently limiting output due to RAM issues.
+    
+#     #test-time dropout
+#     func = K.function(model.inputs + [K.learning_phase()], model.outputs)
+#     pred_ensemble = np.array([np.asarray(func([X] + [1.]), dtype=np.float32).squeeze() for _ in
+#                               range(number_of_forecasts)])
+    
+    #For Full Data
+    #NOTE: Always better to append in a list, rather than as numpy array. bcoz numpy array equires contiguous memory, so if you keep appending it would be slow. alternatively, pre-allocate an empty array. Current method: append to list--> convert to numpy array-->reshape.
+    func = K.function(model.inputs + [K.learning_phase()], model.outputs)
+    preds = []
+    counter=0
+    for X, y in dg_test:
+        preds.append(np.array([np.asarray(func([X] + [1.]), dtype=np.float32).squeeze() for _ in
+                              range(number_of_forecasts)]))
+        print(counter)
+        counter=counter+1
+    pred_ensemble = np.array(preds)
+    
+    #reshaping. Double CHECK!
+    shp=pred_ensemble.shape
+    pred_ensemble=pred_ensemble.transpose(1,0,2,3,4,5).reshape(shp[1],-1,shp[-3],shp[-2],shp[-1])
```

Comment:
Why do you need to reshape the array?

Also, if it is necessary, it would be cleaner to reshape after converting to xarray using the transpose method. That way you can explicitly use the dimension names. Using indices is always dangerous because it's so easy to make errors.

Reply:
The list appends in these dimensions: (batch_number,number_of_forecasts,batch_size,32,64,2)
i.e., after 5th iteration it would be like (5,100,32,32,64,2), but I want (100,160,32,64,2), hence I needed to reshape.

there are other problems with using list too, like last batch size is not same and hence converting to numpy array is not going right. maybe I could try to create an empty numpy array. whenever you get time, please check through the code on the PC. I haven't committed to git yet since its still wrong.

Reply:
Ok, I see the problem. Once you are done, I will have a look whether there is a nicer way of doing this but we probably need some transposing.

Reply:
I stopped the iteration for right now, since it didn't finish even in 3 hours. I'll leave it running tonight